moveit_core/planning_interface/include/moveit/planning_interface/planning_request.h

```diff
@@ -36,12 +36,22 @@
 
 #pragma once
 
+#include <Eigen/Core>
 #include <moveit_msgs/msg/motion_plan_request.hpp>
 
 namespace planning_interface
 {
-// for now this is just a typedef
 
-typedef moveit_msgs::msg::MotionPlanRequest MotionPlanRequest;
+/** \brief Definition for a cost function to measure the cost of a single state during motion planning */
+using StateCostFn = std::function<double(const Eigen::VectorXd& state_vector)>;
+
+struct MotionPlanRequest : moveit_msgs::msg::MotionPlanRequest
+{
+  MotionPlanRequest(moveit_msgs::msg::MotionPlanRequest request_msg = moveit_msgs::msg::MotionPlanRequest(),
+                    planning_interface::StateCostFn state_cost_function = nullptr);
+  planning_interface::StateCostFn state_cost_function = nullptr;
```

Comment:
I guess we need this MotionPlanRequest specialization because the message doesn't have a way to encode cost functions? Does that mean that we can use cost functions only in the C++ interface, but not through messages?
I'm sure I'm missing a lot of context here, but is there a plan to add cost functions to messages, maybe as an enum with pre-defined options?

Reply:
> because the message doesn't have a way to encode cost functions?

Yes :disappointed:, but we had the idea with the pre-defined options too but refrained from adding it for now because we first need to evaluate a set of useful default functions. Until then this is only usable with the cpp interface